### DISCUSSION TITLE
Scanner RPC Design

### DISCUSSION THREAD

**Discussion by mpguerra** [ZcashFoundation/zebra#8156] [Design]
Let's start brainstorming some of the RPC methods that we may want to implement for the Zebra Blockchain scanner

**Comment by mpguerra** [comment]
During yesterday's Zebra team sync we came up with the following list of potential RPC methods we may want to implement:

- RPC server 
    - Get Info
    - Authenticate server 
- Key Management
    - Register a new viewing key
    - Delete a previously registered viewing key
- User/Account Management
    - Register a new account/user
    - Log in
    - Refresh token
- Results
    - Retrieve results for a viewing key
    - Stop/Cancel scan for key

**Reply by arya2** [reply]
The username/password authentication seems redundant if we're checking the keys, it may be enough to have a few methods like:
- `start_scanning(keys: Vec<(Key | KeyHash, Option<ScanJobInfo>)>) -> Vec<KeyHash>`
- `get_transactions(keys: Vec<Key | KeyHash>) -> Vec<(Index, KeyHash, Vec<Transaction>)>`
- `stop_scanning(keys: Vec<(Key | KeyHash, should_delete_key: bool, should_get_transactions: bool)>) -> Vec<(Index, KeyHash, was_deleted: bool, Option<Vec<Transaction>>)>`

with an interceptor (middleware) or a function for either:
- looking up key hashes in the database then passing the keys to method handlers, or
- validating/hashing keys, then passing the key hashes to the method handlers.

and optionally:
- `get_info() -> ScannerInfo`

Clients _should_ be able to authenticate the server if they already have its public key (or a shorter hash of its public key) during the handshake for any request. I'm not yet sure how to do this with tonic/rustls in practice.

**Reply by upbqdn** [reply]
> The username/password authentication seems redundant if we're checking the keys, [...]

If the RPC interface is public, we need to address the following use case:

One user submits a scanning key that others might have, for example, the one from ZECpages. Another user then attempts to delete the key. If Zebra deletes the key and the scanning results, the first user loses their results unexpectedly.

The most natural solution seems to be the "username+(some secret)" model, where scanning keys and scanning results are kept separate for each user.

**Reply by upbqdn** [reply]
There doesn't seem to be any suitable framework for managing client authentication we could use. Tonic supports client certs, but that seems like a lot of work for clients. If we don't use client certs, it looks like we'll need to roll our own management of client secrets.

**Reply by mpguerra** [reply]
That's a good point about the ZECpages key but I feel like that's a special example. In general, you shouldn't be able to delete a key unless you can prove ownership of the key.

**Reply by upbqdn** [reply]
I think I don't follow the proof of the ownership of the key then. Would such proofs prevent the scenario where users try to delete publicly known scanning keys?

**Reply by arya2** [reply]
> Would such proofs prevent the scenario where users try to delete publicly known scanning keys?

I don't think a proof of knowledge for the keys would anyway, you're right - it still needs a username for authorizing access to public viewing keys.

**Reply by mpguerra** [reply]
Let's keep it simple for now and forget about public keys and usernames/passwords

**Reply by upbqdn** [reply]
I think not having support for users doesn't block anything anyway. It should be possible to add that as the last bit. Are we even sure the RPCs will need to be public?

**Reply by arya2** [reply]
> Let's keep it simple for now and forget about public keys and usernames/passwords

Adding usernames may be trivial, otherwise, I agree it should be left until the next version and will exclude it.

> I think not having support for users doesn't block anything anyway. It should be possible to add that as the last bit.

Yeah, I don't think it'll block anything either. 

> Are we even sure the RPCs will need to be public?

I don't think they need to be public for exchanges, though I am hoping some individual users will run their own Zebra nodes at home to use with their mobile wallets.

They may need to be public later if Zcash adopts detection keys.

**Comment by upbqdn** [comment]
I thought of a trivial interface that would, apart from user management, essentially consist of only two RPCs:

- `scan(Vec<Key>, Option<HeightInterval>) -> Stream<(TXID, Key)>`
- `clear_cache_for(Vec<Key>)`

The persistent storage would be transparent for clients, meaning that Zebra would actually perform the scanning only if there were no results in the storage. Otherwise, it would return cached results from the storage.

**Reply by upbqdn** [reply]
But now I'm thinking this might not work for clients that want to initiate the scanning, then disconnect from the server, and fetch the results at a later time.

**Reply by arya2** [reply]
> scan(Vec<Key>, Option<HeightInterval>) -> Stream<(TXID, Key)>

> But now I'm thinking this might not work for clients that want to initiate the scanning, then disconnect from the server, and fetch the results at a later time.

I think this push interface works. If the client disconnects from the server, it can pick up where it left off when it reconnects to fetch the results.

Possible Design:

- start a scanner task
  - `try_send` results to a bounded channel for a small buffer
  - store results in the database if the channel is closed or full
- in the connection handler with the channel receiver:
  - check the database for any cached results since the last disconnection
  - try sending results to the client
  - delete results once the client has successfully received them
  - check the channel for any new results
  - drop results once the client has successfully received them
  - drain the channel to the database once the connection is dropped

Some clients may want to avoid streaming the results for simplicity, and I'm not sure if Tor works with http/2 or http/1.1 sockets, but this should work fine in mobile environments where background tasks are unreliable, and we could add another method later that works with regular http/1.1 requests.

Lightwalletd uses streams, I think it would be okay to provide only that interface for the first version.

**Reply by upbqdn** [reply]
To make the interface more specific:

- Passing an empty vec to `clear_cache_for` would implicitly remove all scanning keys and corresponding results from the storage.
- I don't understand the frameworks we'll use well enough, so I don't yet know how the clients will distinguish between _"scanning finished: no results for the submitted scanning keys"_ and _"no results, but scanning didn't finish yet"_. 

If we go with these two RPCs, I don't see any other ones we would need.

**Comment by mpguerra** [comment]
This is the latest design:
![zebra-grpc](https://github.com/ZcashFoundation/zebra/assets/1311133/93831807-13a2-4400-a080-909122caf06d)

Green items are already implemented
Lilac/Light blue items are work for current iterations/milestone 
Dark Blue items are future, currently unplanned work

**Comment by mpguerra** [comment]
We're not going to do any further work on the scanner
